src/hid/joystick_sfml_keyboard.cc

```diff
@@ -0,0 +1,97 @@
+// BoB robotics includes
+#include "hid/joystick_sfml_keyboard.h"
+
+template<class T, class Func>
+bool runIfFound(const std::map<sf::Keyboard::Key, T> m,
+                sf::Keyboard::Key key,
+                Func f)
+{
+    auto iter = m.find(key);
+    if (iter == m.end()) { // not present
+        return false;
+    }
+
+    // Run f with the value corresponding to key
+    f(iter->second);
+    return true;
+}
+
+namespace BoBRobotics {
+namespace HID {
+
+const std::map<sf::Keyboard::Key, JAxis> JoystickSFMLKeyboard::NegativeAxisKeys = {
+    { sf::Keyboard::Key::W, JAxis::LeftStickVertical },
+    { sf::Keyboard::Key::A, JAxis::LeftStickHorizontal },
+    { sf::Keyboard::Key::Up, JAxis::RightStickVertical },
+    { sf::Keyboard::Key::Left, JAxis::RightStickHorizontal },
+    { sf::Keyboard::Key::PageUp, JAxis::RightTrigger }
+};
+const std::map<sf::Keyboard::Key, JAxis> JoystickSFMLKeyboard::PositiveAxisKeys = {
+    { sf::Keyboard::Key::S, JAxis::LeftStickVertical },
+    { sf::Keyboard::Key::D, JAxis::LeftStickHorizontal },
+    { sf::Keyboard::Key::Down, JAxis::RightStickVertical },
+    { sf::Keyboard::Key::Right, JAxis::RightStickHorizontal },
+    { sf::Keyboard::Key::PageDown, JAxis::RightTrigger }
+};
+const std::map<sf::Keyboard::Key, JButton> JoystickSFMLKeyboard::ButtonKeys = {
+    { sf::Keyboard::Key::Num1, JButton::A },
+    { sf::Keyboard::Key::Num2, JButton::B },
+    { sf::Keyboard::Key::Num3, JButton::X },
+    { sf::Keyboard::Key::Num4, JButton::Y }
+};
+
+//------------------------------------------------------------------------
+// BoBRobotics::HID::JoystickSFMLKeyboard
+//------------------------------------------------------------------------
+JoystickSFMLKeyboard::JoystickSFMLKeyboard(sf::Window &window)
+  : JoystickBase(0.0f), m_Window(window)
+{}
+
+//------------------------------------------------------------------------
+bool JoystickSFMLKeyboard::updateState()
+{
+    sf::Event event;
+    if (!m_Window.pollEvent(event)) {
```

Comment:
Hmm, I think this is problematic, it looks like (I haven't actually used SFML before in anger) this function is supposed to be called in your main loop so:

1. Multiple events will need to be processed each update
2. They may be required by some other sub-system

I think what you need is to use ``sf::Keyboard::isKeyPressed``. So many something like:

```c++
for(auto key : ButtonKeys) {
    // Determines whether key is now down (I THINK this must what this API actually returns)
    const bool nowDown = sf::Keyboard::isKeyPressed(key.first);
    
    // If state has changed
    if(previouslyDown != isDown(key.second)) {
        if(previouslyDown) {
           setReleased(key.second, false);
        }
        else {
            setPressed(key.second, false);
        }
    }
}
```
You could also then make it ``ButtonKeys`` a fixed size array ``JButton::Max`` although, this always end  up being ugly with class enums (which is one of the things that really irritates me with C++11)

Reply:
That is a good point. I did think about this: for example, our ``Viz::SFMLWorld`` class won't work in conjunction with this code because it also calls ``pollEvents()``. Do you think there's a chance it might miss keyboard events this way though?

Reply:
Yeah, as far as I can tell, ``pollEvents`` should only be called in the outer loop of the application and events dispatched from there - if you've ever had the misfortune of ye olde win 32 API programming it works in a similar way.

Potentially but, assuming we don't lazily do loads of stuff in the main thread (which, admittedly, I often do) it would require users doing some serious sub-ms key-bashing :smile:

Reply:
Good point. I guess if we're doing too much on the main thread then that's an issue by itself. Could always run ``joystick.runInBackground()`` as a hack in that case.